File: android/app/src/main/java/com/oru/radio/ControlMessage.kt
package com.oru.radio

import org.json.JSONException
import org.json.JSONObject

/** The reliable BYTES control channel of spec section 7. */
sealed class ControlMessage {
    data class Hello(val version: Int) : ControlMessage()
    data class TxStart(val streamId: String) : ControlMessage()
    data class TxStop(val streamId: String) : ControlMessage()
}

/**
 * JSON on the wire, exactly the shapes in spec section 7. iOS and the TypeScript layer
 * implement the same three shapes; key order is irrelevant, key names are not.
 */
object ControlMessageCodec {

    fun encode(message: ControlMessage): ByteArray {
        val json = when (message) {
            is ControlMessage.Hello ->
                JSONObject().put("type", "hello").put("version", message.version)
            is ControlMessage.TxStart ->
                JSONObject().put("type", "tx-start").put("streamId", message.streamId)
            is ControlMessage.TxStop ->
                JSONObject().put("type", "tx-stop").put("streamId", message.streamId)
        }
        return json.toString().toByteArray(Charsets.UTF_8)
    }

    /** Returns null for anything this version cannot use; a peer never crashes us. */
    fun decode(bytes: ByteArray): ControlMessage? = try {
        val json = JSONObject(String(bytes, Charsets.UTF_8))
        when (json.optString("type")) {
            "hello" -> if (json.get("version") is Int) ControlMessage.Hello(json.getInt("version")) else null
            "tx-start" -> streamId(json)?.let(ControlMessage::TxStart)
            "tx-stop" -> streamId(json)?.let(ControlMessage::TxStop)
            else -> null
        }
    } catch (e: JSONException) {
        null
    }

    private fun streamId(json: JSONObject): String? =
        json.optString("streamId").takeIf { it.isNotEmpty() }
}

File: ios/Radio/Sources/RadioKit/ControlMessage.swift
import Foundation

/// The reliable BYTES control protocol of spec section 7. The JSON shape is a
/// cross-platform contract: the Android engine and the TypeScript codec parse the
/// exact same three objects.
public enum ControlMessage: Equatable {
    case hello(version: Int)
    case txStart(streamId: String)
    case txStop(streamId: String)

    private enum Wire {
        static let type = "type"
        static let version = "version"
        static let streamId = "streamId"
        static let hello = "hello"
        static let txStart = "tx-start"
        static let txStop = "tx-stop"
    }

    public func encoded() -> Data {
        let object: [String: Any]
        switch self {
        case let .hello(version):
            object = [Wire.type: Wire.hello, Wire.version: version]
        case let .txStart(streamId):
            object = [Wire.type: Wire.txStart, Wire.streamId: streamId]
        case let .txStop(streamId):
            object = [Wire.type: Wire.txStop, Wire.streamId: streamId]
        }
        // The three shapes above are always serialisable, so the failure branch is
        // unreachable; an empty payload is decoded as nil by the peer either way.
        return (try? JSONSerialization.data(withJSONObject: object)) ?? Data()
    }

    public static func decode(_ data: Data) -> ControlMessage? {
        guard
            let object = try? JSONSerialization.jsonObject(with: data),
            let json = object as? [String: Any],
            let type = json[Wire.type] as? String
        else {
            return nil
        }

        switch type {
        case Wire.hello:
            guard let version = json[Wire.version] as? Int else { return nil }
            return .hello(version: version)
        case Wire.txStart:
            guard let streamId = json[Wire.streamId] as? String else { return nil }
            return .txStart(streamId: streamId)
        case Wire.txStop:
            guard let streamId = json[Wire.streamId] as? String else { return nil }
            return .txStop(streamId: streamId)
        default:
            return nil
        }
    }
}
